man/uprecords.1
.TH UPRECORDS 1 "June 20, 2004"
.nh

.SH NAME
uprecords \- provides uptime records

.SH SYNOPSIS
.B uprecords
[ \fB-habBcfkKMsv\fP ] [ \fB-i\fP <interval> ] [ \fB-m\fP <count> ]

.SH DESCRIPTION
This manual page documents the
.B uprecords
program.
.PP
.B uprecords
is a program that provides record statistics from the
.B uptimed(8)
program.

.SH OPTIONS
.TP
.B \-h
Show usage
.TP
.B \-a
Do not print ANSI codes
.TP
.B \-b -B
Sort by time of bootup, oldest entry first. -B reverses. Option implies \-s.
Do not print ANSI codes
.TP
.B \-c
Do not show current entry if not in top entries
.TP
.B \-f
Run continuously in a loop
.TP
.B \-k -K
Sort by kernel/system name. -K reverses. Option implies \-s.
.TP
.B \-M
Show next milestone
.TP
.B \-s
Do not print extra statistics
.TP
\fB\-i\fP <interval>
Use <interval> seconds for loop instead of 5, implies -f
.TP
\fB\-m\fP <count>
Show a maximum of top COUNT entries instead of 10
.TP
.B \-v
Show version information

.SH SEE ALSO
.BR uptimed (8)
.br

.SH AUTHOR
Rob Kaper <cap@capsi.com>.

This manual page was written by Alan Ford <alan@whirlnet.co.uk>, for the
Debian GNU/Linux system (but may be used by others). It was last modified
for
.B uptimed
version 0.3.7.


NEWS
0.3.7
-----

- it should now build correctly
  (thanks to Michael Ploujnikov <ploujj@gmail.com>)

0.3.6
-----

- patch by Thibaut Varene <t-bone@parisc-linux.org>: add darwin support
  (define it as BSD system).

0.3.5
-----

- version numbers should now be correct.

0.3.4
-----

- patch by Radek Podgorny <radek@podgorny.cz>: add a "don't fork" option.
- this is the first release under maintainership of Radek Podgorny.

0.3.3
-----

- patch by Brett: allow sorting uprecords output.

0.3.2
-----

- change SYSMAX to 24 characters to accomodate FreeBSD 4.8-RELEASE-p10 and
  up.
- remove bootid functionality for BSD: it is not needed. Updated
  documentation.

0.3.1
-----

- accepted patch from Radim Kolar <hsn@netmag.cz> for bug in save_records
  that could cause data loss on a system crash with the records file opened
  for writing.

2002-11-05 (uptimed-0.3.0)
----------

- fixed log_threshold
- added mail_threshold

2002-05-13 (cap) (uptimed-0.2.0 release)
----------

- various documentation updates

2002-03-16 (generica)
----------

- Add etc dir
- Move uptimed.conf-dist to etc dir
- Create template rc.uptimed in etc dir
- Update spec file to use rc.uptimed

2001-12-21 (generica)
----------

- Update spec file to deal with sample-cgi dir properly,
  not overwrite /etc/uptimed.conf, warn user to add/remove
  lines to rc.(sysinit,local), create /var/spool/uptimed

2001-12-21 (generica)
----------

- Run 'libtoolize' in autogen.sh

2001-12-20 (generica)
----------

- Add rpm spec file

2001-12-19 (cap)
----------

- Added hostname to e-mail confirmation message and beautified the contents
  of the e-mail.

2001-12-17 (generica)
----------

- New directory (libuptimed)
- Create libuptimed.a and libuptimed.so using libtool
- Add uptimed.lsm to be included in `make dist`

2001-12-12 (generica)
----------

- Output location of binary when printing usage/help
- Small typos in INSTALL, README, uptimed.8, uprecords.1 and urec.c

2001-12-05 (cap) (uptimed-0.1.7 release)
----------

- Change to getopt (generica)
- Include local copy of getopt.[ch] for systems without (generica)
- Small updates in README, INSTALL
- Updated deprecated capsi.cx references
- Replaced strcat/sprintf with their respective 'n' functions
- uprecords -M works again (fixes Debian bug #107492)
- no longer overwrite uptimed.conf upon installation

2001-07-30 (generica)
----------

- Added autoconf/automake support
- New directory structure (src/man)

2001-07-27 (cap)
----------

- new ChangeLog syntax ;-)
- Updated TODO
- moved uptimed code to CVS

0.1.7 - not released
====================

* More checks to prevent bogus mail reported fixed in 0.1.5.


0.1.6 - Nov/06/2000
===================

* Manual pages.
* Easier configuration of milestones in config file.
* A few little bugfixes.

0.1.5 - Mar/30/2000
===================

* Bugfix for faulty interpretation of 0 (unlimited) entries in uptimed and
  uprecords.
* Bugfix against startup crashes with corrupted startup files by adding a
  syntax check on input lines.
* Phoney entries in old records file (version 0.03 and earlier) are now
  discarded by uptimed.
* Bugfix to eliminate an odd congratulations mail on startup.
* Improved 'make install' in Makefile, no longer overwrites existing
  configuration but instead install the template as /etc/uptimed.conf.dist

0.1.4 - Mar/09/2000
===================

* Linux uptime now retrieved from sysinfo() instead of /proc/uptime. (patch
  from Vitezslav Samel <vsamel@seznam.cz>)
* Changed default email address in config from my own (stupid, stupid!) to
  root@localhost.
* Configurable milestones.
* Moved all .h declarations to header files.
* New option -M to show next milestone in uprecords.

0.1.3 - Feb/28/2000
===================

* Partial HP/UX support. (patches from Helge Deller <deller@gmx.de>)

0.1.2 - Dec/18/1999
===================

* Fixed -s and -v options in uprecords.
* Fixed parsing of email address.
* Added partial support for *BSD and Solaris.
* Fixed wrong output in email.
* Added sendmail location to configuration file.

0.1.1 - Dec/9/1999
==================

* Added commandline options to uptimed and uprecords. (see -?)
* Option not to show ANSI codes in uprecords. (for inclusion into .signature
  files, for example)
* Option not to show current entry, statistics.
* Fixed instructions in INSTALL file.
* Fixed formatting bug in uprecords for too long system info.
* Option to mail records/milestones.

0.1.0 - Dec/6/1999
==================

Basically a complete rewrite of the code:
* Created urec struct to put records in and using linked list instead of
  ugly static arrays.
* Files now stored in /var/spool/uptimed/
* Made 'make uninstall' entry in Makefile.
* Events such as moving up in the list, breaking the record or reaching a
  milestone get reported in syslog.
* Compiles clean with -Wall -pedantic -ansi on my system :-)
* Update interval and maximum amount of log entries are now configurable
  from configuration file /etc/uptimed.conf
* Rewrote INSTALL and README files.
* Added CGI functionality to uprecords and wrote INSTALL.cgi
* Catching signals SIGHUP/SIGTERM to exit gracefully upon them.

0.03 - 01/14/1999
=================

* Added countdowns for milestones in uprecords.
* Bugfixes.
* Added -f option for uprecords. Runs the program in a loop and doesn't
  exit. Handy for spare monitors and/or terminals.
* Added system recognition. Lost backwards compatibility in logfile. See
  README.upgrade

0.02 - 01/04/1999
=================

* Rewrote the code that is used to identify a unique boot id. See the
  INSTALL file for details on the method I have chosen, how it works and why
  I have chosen it.

0.01 - 12/13/1998
=================

* First release.


src/uprecords.c
/*
uptimed - Copyright (c) 1998-2004 Rob Kaper <rob@unixcode.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; version 2 dated June, 1991.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program;  if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave., Cambridge, MA 02139, USA.
*/

#include "../config.h"
#include "uprecords.h"

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define SYSWIDTH 24
#define DOWNWIDTH 20

#if defined(__ANDROID__) || defined(PLATFORM_AIX)
extern Urec *u_current;
#else
Urec	*u_current;
#endif
time_t	first, prev, tenth, second;
int		runas_cgi=0, show_max=10, show_milestone=0, layout=PRE, show_downtime=0, run_loop=0, update_interval=5;
int		sort_by=0, no_ansi=0, no_stats=0, no_current=0, wide_out=0;

int main(int argc, char *argv[])
{
	/* Read config file. */
	read_config();

	/* Check if we are being run as CGI program. */
	if (strstr(argv[0], ".cgi"))
	{
		runas_cgi=1;

		/* Print content-type header. */
		printf("Content-type: text/html\n\n");

		/* Read CGI config file. */
		read_config_cgi();
	}

	scan_args(argc, argv);

#ifdef PLATFORM_HPUX
	no_ansi=1;
#endif

	if (update_interval<1)
		update_interval=1;

	if (show_max<0)
		show_max=1;

	/* Print header file. */
	if (runas_cgi)
		cat("/etc/uprecords-cgi/uprecords.header");

	/* Read current uptime and entries from logfile. */
	u_current=add_urec(read_uptime(), readbootid(), read_sysinfo());
	read_records(u_current->btime);

	if (!runas_cgi && run_loop)
	{
		while(1)
		{
			/* Update current uptime. */
			u_current->utime=read_uptime();

			/* Current uptime might move up in the list. */
			if ((prev && prev - u_current->utime < 0) || (first && first - u_current->utime < 0))
			{
				moveup();
				first=0;
			}
			displayrecords(1);
			sleep(update_interval);
		}
	}
	else displayrecords(0);

	/* Print footer file. */
	if (runas_cgi)
		cat("/etc/uprecords-cgi/uprecords.footer");

	return 0;
}

void displayrecords(int cls)
{
	Urec *u, *uprev=NULL;
	time_t since, now, tmp, totalutime = 0, totaldtime = 0;
	float availability;
	int	i=0, currentdone=0;

	now=time(0);

	/* Open output for CGI. */
	if (runas_cgi)
	{
		if (layout==TABLE)
			printf("<table border=1>\n");
		else if (layout==LIST)
			printf("<ol>\n");
		else
			printf("<pre>\n");
	}
	/* Clear screen. */
	else if (cls && !no_ansi)
		printf(CLS);

	if (layout==PRE)
	{
		if (!show_downtime) {
			printf("   %3s%21s | %-*s %*s\n", "#", "Uptime", SYSWIDTH, "System", TIMEMAX, "Boot up");
		} else {
			printf("   %3s%21s | %*s %*s\n", "#", "Uptime", DOWNWIDTH, "Last downtime", DOWNTIMEMAXWIDTH, "Boot up");
		}
		print_line();
	}
	else if (runas_cgi && layout==TABLE)
	{
		printf("<th>%s</th>\n", "Position");
		printf("<th>%s</th>\n", "Uptime");
		if (!show_downtime) {
			printf("<th>%s</th>\n", "System");
		} else {
			printf("<th>%s</th>\n", "Last downtime");
		}
		printf("<th>%s</th>\n", "Boot up");
	}

	urec_list = sort_urec(urec_list, sort_by);

	for ( u=urec_list; u ; u=u->next )
	{
		if (++i<=show_max || show_max==0)
		{
			if (u==u_current)
			{
				if (!show_downtime) {
					print_entry(u->utime, u->sys, u->btime, "-> ", i, 1);
				} else {
					print_downtime_entry(u->utime, u->dtime, u->btime, "-> ", i, 1);
				}
				currentdone++;
				if (uprev) prev=uprev->utime;
				else
				{
 					prev=0; first=0; tenth=0; second=0;
				}
			}
			else
			{
				if (!show_downtime) {
					print_entry(u->utime, u->sys, u->btime, "   ", i, 0);
				} else {
					print_downtime_entry(u->utime, u->dtime, u->btime, "   ", i, 0);
				}
				if (i==1) first=u->utime;
 				if (i==2) second=u->utime;
 				if (i==10) tenth=u->utime;
			}
		}
		else if (u==u_current)
		{
			if (uprev) prev=uprev->utime;
			break;
		}
		uprev=u;
	}

	if (runas_cgi && layout==LIST)
	{
		printf("</ol>\n");
		printf("<ul>\n");
	}

	if (!no_current && !currentdone)
	{
		if (layout==PRE)
			print_line();

		print_entry(u_current->utime, u_current->sys, u_current->btime, "-> ", i, 1);
		if (uprev) prev=uprev->utime;
	}

	if (!no_stats)
	{
		if (layout==PRE)
			print_line();

		if (prev && prev!=first)
		{
			tmp=now + prev - u_current->utime;
			print_entry(prev - u_current->utime + 1, "at", tmp, "1up in", 0, 0);
		}
		if (tenth && prev<=tenth && prev!=first)
		{
			tmp=now + tenth - u_current->utime;
			print_entry(tenth - u_current->utime + 1, "at", tmp, "t10 in", 0, 0);
		}
		if (first)
		{
			tmp=now + first - u_current->utime;
			print_entry(first - u_current->utime + 1, "at", tmp, "no1 in", 0, 0);
		} else {
			tmp=now + second - u_current->utime;
			print_entry(u_current->utime - second - 1, "since", tmp, "NewRec", 0, 0);
		}
		if (show_milestone)
		{
			Milestone *m;

			m = find_next_milestone(u_current->utime);
			if (m!=NULL)
			{
				tmp=now + m->time - u_current->utime;
				print_entry(m->time - u_current->utime + 1, m->desc, tmp, "mst in", 0, 0);
			}
		}

		/* Printing total uptime and downtime. */
		for (u = urec_list; u; u = u->next){
			if (u->dtime == 0) {
				since = u->btime;
			}
			totaldtime += u->dtime;
			totalutime += u->utime;
		}

		print_entry(totalutime, "since", since, "up", 0, 0);
		print_entry(totaldtime, "since", since, "down", 0, 0);

		/* Printing availability. */
		availability = (float)totalutime / (float)(totalutime + totaldtime) * 100;
		print_availability(availability, since);
	}

	/* End output for CGI. */
	if (runas_cgi)
	{
		if (layout==TABLE)
			printf("</table>\n");
		else if (layout==LIST)
			printf("</ul>\n");
		else
			printf("</pre>\n");

		printf("<small><a href=\"https://github.com/rpodgorny/uptimed\">uptimed</a> by Rob Kaper (<a href=\"mailto:rob@unixcode.org\">rob@unixcode.org</a>) - currently maintained by Radek Podgorny (<a href=\"mailto:radek@podgorny.cz\">radek@podgorny.cz</a>)</small>\n");
	}
}

void read_config(void)
{
	FILE *f;
	char str[256];
	time_t milestone_time;
	char *milestone_str;

	f=fopen(FILE_CONFIG, "r");
	if (!f) return;

	fgets(str, sizeof(str), f);
	while (!feof(f))
	{
		if (!strncmp(str, "MILESTONE",  9))
		{
			char *cp = strtok(str+10, ":");
			milestone_time=scantime(cp);
			milestone_str=strtok(NULL, "\n");
			add_milestone(milestone_time, milestone_str);
		}
		fgets(str, sizeof(str), f);
	}
}

void read_config_cgi(void)
{
	FILE *f;
	char str[256];

	f=fopen("/etc/uprecords-cgi/uprecords.conf", "r");
	if (!f)
		return;

	fgets(str, sizeof(str), f);
	while(!feof(f))
	{
		if (!strncmp(str, "LAYOUT", 6))
		{
			if (!strncmp(str+7, "table", 5))
				layout=TABLE;
			else if (!strncmp(str+7, "list", 4))
				layout=LIST;
			else
				layout=PRE;
		}
		else if (!strncmp(str, "SHOW_MAX", 8))
			show_max=atoi(str+9);
		else if (!strncmp(str, "TYPE", 4))
		{
			if (!strncmp(str+5, "system", 6))
				show_downtime = 0;
			else if (!strncmp(str+5, "downtime", 8))
				show_downtime = 1;
		}
		fgets(str, sizeof(str), f);
	}
	fclose(f);
}

void print_entry(time_t utime, char *sys, time_t btime, char *ident, int pos, int hilite)
{
	char *bold = BOLD, *plain = PLAIN, *current = "";
	char *ctimec = NULL;

	if (runas_cgi)
	{
		bold = "<b>";
		plain = "</b>";

		if (hilite)
			current = " (current)";

		if (!strcmp(ident, "-> "))
			ident="-&gt; ";

		if (layout!=PRE)
		{
			if (!strcmp(ident, "1up in"))
				ident="One up in";
			else if (!strcmp(ident, "no1 in"))
				ident="Number one in";
		}
	}

	if (!hilite || no_ansi)
	{
		bold = "";
		plain = "";
	}

	switch(layout)
	{
		case TABLE:
			printf("<tr>\n");
			if (pos)
				printf("<td>%s%d%s%s</td>\n", bold, pos, plain, current);
			else
				printf("<td>%s</td>\n", ident);

			printf("<td>%s%s%s</td>\n", bold, time2uptime(utime), plain);
			printf("<td>%s%s%s</td>\n", bold, sys, plain);
			printf("<td>%s%s%s</td>\n", bold, ctime(&btime), plain);

			printf("</tr>\n");
			break;
		case LIST:
			if (pos>show_max)
				printf("<li>%s%s, %s, %s%s%s (position %d)", bold, time2uptime(utime), sys, ctime(&btime), plain, current, pos);
			else if (pos)
				printf("<li>%s%s, %s, %s%s%s", bold, time2uptime(utime), sys, ctime(&btime), plain, current);
			else
				printf("<li>%s%s %s %s %s%s", bold, ident, time2uptime(utime), sys, ctime(&btime), plain);
			break;
		default:
			if((ctimec = ctime(&btime)))
				ctimec[TIMEMAX-1] = '\0';	/* erase the ending '\n' */
			if (!wide_out && strlen(sys) > SYSWIDTH)
				sys[SYSWIDTH] = '\0';		/* truncate for 80 cols */
			if (pos)
				printf("%s%3s%3d%21s %s|%s %-*s %*s%s\n", bold, ident, pos, time2uptime(utime), plain, bold, SYSWIDTH, sys, TIMEMAX, ctimec, plain);
			else
				printf("%s%6s%21s %s|%s %-*s %*s%s\n", bold, ident, time2uptime(utime), plain, bold, SYSWIDTH, sys, TIMEMAX, ctimec, plain);
	}
}

void print_downtime_entry(time_t utime, time_t dtime, time_t btime, char *ident, int pos, int hilite)
{
	char *bold = BOLD, *plain = PLAIN, *current = "";
	char *ctimec = NULL;
	char timebuf[20] = "", dtimebuf[20] = "";

	strcpy(timebuf, time2uptime(utime));
	strcpy(dtimebuf, time2uptime(dtime));

	if (runas_cgi)
	{
		bold = "<b>";
		plain = "</b>";

		if (hilite)
			current = " (current)";

		if (!strcmp(ident, "-> "))
			ident="-&gt; ";
	}

	if (!hilite || no_ansi)
	{
		bold = "";
		plain = "";
	}

	switch(layout)
	{
		case TABLE:
			printf("<tr>\n");
			if (pos)
				printf("<td>%s%d%s%s</td>\n", bold, pos, plain, current);
			else
				printf("<td>%s</td>\n", ident);

			printf("<td>%s%s%s</td>\n", bold, timebuf, plain);
			printf("<td>%s%s%s</td>\n", bold, dtimebuf, plain);
			printf("<td>%s%s%s</td>\n", bold, ctime(&btime), plain);

			printf("</tr>\n");
			break;
		case LIST:
			if (pos>show_max)
				printf("<li>%s%s, %s, %s%s%s (position %d)", bold, timebuf, dtimebuf, ctime(&btime), plain, current, pos);
			else if (pos)
				printf("<li>%s%s, %s, %s%s%s", bold, timebuf, dtimebuf, ctime(&btime), plain, current);
			else
				printf("<li>%s%s %s at %s%s", bold, ident, timebuf, ctime(&btime), plain);
			break;
		default:
			if((ctimec = ctime(&btime)))
				ctimec[TIMEMAX-1] = '\0';	/* erase the ending '\n' */
			if (pos) {
				printf("%s%3s%3d%21s %s|%s %*s %*s%s\n", bold, ident, pos, timebuf, plain, bold, DOWNWIDTH, dtimebuf, DOWNTIMEMAXWIDTH, ctimec, plain);
			} else {
				printf("%s%6s%21s %s|%s %*s %*s%s\n", bold, ident, timebuf, plain, bold, DOWNWIDTH, dtimebuf, DOWNTIMEMAXWIDTH, ctimec, plain);
			}
	}
}

void print_availability(float percent, time_t since)
{
	char *ctimec = NULL, *msg = "%up";

	if (runas_cgi) {
		if (layout!=PRE) {
			msg = "Availability (%)";
		}
	}

	switch(layout)
	{
		case TABLE:
			printf("<tr>\n");
			printf("<td>%s</td>\n", msg);
			printf("<td>%.3f</td>\n", percent);
			printf("<td>since</td>\n");
			printf("<td>%s</td>\n", ctime(&since));

			printf("</tr>\n");
			break;
		case LIST:
				printf("<li>%s %.3f since %s", msg, percent, ctime(&since));
			break;
		default:
			if (ctimec = ctime(&since)) {
				ctimec[TIMEMAX-1] = '\0';	/* erase the ending '\n' */
			}
			printf("%6s %20.3f | %-*s %*s\n", msg, percent, SYSWIDTH, "since", TIMEMAX, ctimec);
	}
}

void print_line(void)
{
	printf("----------------------------+---------------------------------------------------\n");
}

void scan_args(int argc, char *argv[])
{
	int i;

	while((i = getopt(argc, argv, "i:m:?hacbdBkKfsMwv")) != EOF)
	{
		switch(i)
		{
				case '?':
				case 'h':
#ifdef __ANDROID__
						fputc('\n', stderr);
#endif
						print_help(argv);
						break;
				case 'a':
						no_ansi++;
						break;
				case 'b':
						sort_by = 1;
						no_stats++;
						break;
				case 'B':
						sort_by = -1;
						no_stats++;
						break;
				case 'k':
						sort_by = 2;
						no_stats++;
						break;
				case 'K':
						sort_by = -2;
						no_stats++;
						break;
				case 'd':
						show_downtime++;
						break;
				case 'c':
						no_current++;
						break;
				case 'f':
						run_loop++;
						break;
				case 's':
						no_stats++;
						break;
				case 'w':
						wide_out++;
						break;
				case 'v':
						print_version();
						break;
				case 'M':
					show_milestone++;
					break;
				case 'i':
					if (isdigit(*optarg))
						update_interval=atoi(optarg);
					else
					{
						printf("%s: option requires argument -- %c\n", argv[0], i);
						print_usage(argv);
					}
					run_loop++;
					break;
				case 'm':
					if (isdigit(*optarg))
						show_max=atoi(optarg);
					else
					{
						printf("%s: option requires argument -- %c\n", argv[0], i);
						print_usage(argv);
					}
					break;
		}
	}
}

void print_usage(char *argv[])
{
	printf("usage: %s [-hacdfMswv] [-i interval] [-m count]\n", argv[0]);
	exit(1);
}

void print_help(char *argv[])
{
	printf("usage: %s [OPTION]...\n\n", argv[0]);
	printf("  -h             this help\n");
	printf("  -a             do not print ansi codes\n");
	printf("  -b             sort by boottime\n");
	printf("  -B             reverse sort by boottime\n");
	printf("  -k             sort by sysinfo\n");
	printf("  -K             reverse sort by sysinfo\n");
	printf("  -d             print downtime seen before every uptimes instead of system\n");
	printf("  -c             do not show current entry if not in top entries\n");
	printf("  -f             run continuously in a loop\n");
	printf("  -s             do not print extra statistics\n");
	printf("  -w             wide output (more than 80 cols per line)\n");
	printf("  -i INTERVAL    use INTERVAL seconds for loop instead of 5, implies -f\n");
	printf("  -m COUNT       show a maximum of top COUNT entries instead of 10\n");
	printf("  -M             show next milestone\n");
	printf("  -v             version information\n");
	exit(0);
}

void print_version(void)
{
	printf("uprecords " VERSION " by Rob Kaper <rob@unixcode.org>\n");
	printf("enhanced and maintained by Radek Podgorny <radek@podgorny.cz>\n");
	exit(0);
}
